galley/pkg/config/processing/snapshotter/strategy/immediate.go

```diff
@@ -0,0 +1,57 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package strategy
+
+import (
+	"sync/atomic"
+
+	"istio.io/istio/galley/pkg/runtime/monitoring"
+)
+
+// Immediate is a snapshotting strategy for creating snapshots immediately, as events arrive.
+type Immediate struct {
+	handler atomic.Value
+}
+
+func sentinelOnSnapshot() {}
+
+var _ Instance = &Immediate{}
+
+// NewImmediate returns a new Immediate.
+func NewImmediate() *Immediate {
+	i := &Immediate{}
+	i.handler.Store(OnSnapshotFn(sentinelOnSnapshot))
+
+	return i
+}
+
+// Start implements processing.Debounce
+func (i *Immediate) Start(handler OnSnapshotFn) {
+	i.handler.Store(handler)
+}
+
+// Stop implements processing.Debounce
+func (i *Immediate) Stop() {
+	i.handler.Store(OnSnapshotFn(sentinelOnSnapshot))
+}
+
+// OnChange implements processing.Debounce
+func (i *Immediate) OnChange() {
+	scope.Debug("Immediate.OnChange")
+	fn := i.handler.Load().(OnSnapshotFn)
```

Comment:
I'm assuming you're using atomic for performance vs mutex.  Have you benchmarked both?

Reply:
I didn't originally, and guessed that it would be a wash at best (as mutex is using atomic compare/swap for its fast path), But out of curiosity I did a quick and dirty test:
```
BenchmarkAtomic/#00-12 	1000000000	         2.54 ns/op	       0 B/op	       0 allocs/op
BenchmarkMutex/#00-12  	100000000	        15.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkRWMutex/#00-12         	100000000	        12.7 ns/op	       0 B/op	       0 allocs/op
```
With this code:
```

type aType struct {
	a atomic.Value
}

func BenchmarkAtomic(b *testing.B) {
	a := &aType{}
	a.a.Store(func() {})

	b.Run("", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			fn := a.a.Load().(func())
			fn()
		}
	})
}

type muType struct {
	mu sync.Mutex
	fn func()
}

func BenchmarkMutex(b *testing.B) {
	m := &muType{
		fn: func() {},
	}

	b.Run("", func(b *testing.B) {

		for i := 0; i < b.N; i++ {
			m.mu.Lock()
			fn := m.fn
			m.mu.Unlock()
			fn()
		}
	})
}


type rwMuType struct {
	mu sync.RWMutex
	fn func()
}

func BenchmarkRWMutex(b *testing.B) {
	m := &rwMuType{
		fn: func() {},
	}

	b.Run("", func(b *testing.B) {

		for i := 0; i < b.N; i++ {
			m.mu.RLock()
			fn := m.fn
			m.mu.RUnlock()
			fn()
		}
	})
}

```

Reply:
Thanks ... Looks good!